### PR TITLE
[IMP] mail: im status icon 

### PR DESCRIPTION
**Pull request by maki-odoo**
[IMP] mail: im status icon 

By this commit, the im status appears in the messaging menu, the sidebar menu and the chatter aside

### CODE REVIEW

File: addons/mail/static/src/thread/message.js [hunk 1]
```diff
@@ -3,6 +3,7 @@
 import { useMessaging } from "../messaging_hook";
 import { RelativeTime } from "./relative_time";
 import { Component, useExternalListener, useRef, useState } from "@odoo/owl";
+import { PartnerImStatus } from "../discuss/partner_im_status";
```
Comment (phenix-factory) [inline addons/mail/static/src/thread/message.js]: Use @mail import instead of relative path :)

File: addons/mail/static/src/messaging_menu/messaging_menu.js [hunk 2]
```diff
@@ -4,6 +4,7 @@ import { Component, useState } from "@odoo/owl";
 import { Dropdown } from "@web/core/dropdown/dropdown";
 import { registry } from "@web/core/registry";
 import { useMessaging } from "../messaging_hook";
+import { PartnerImStatus } from "../discuss/partner_im_status";
```
Comment (phenix-factory) [inline addons/mail/static/src/messaging_menu/messaging_menu.js]: Better use @mail import

File: addons/mail/static/src/messaging.js [hunk 3]
```diff
@@ -263,6 +275,15 @@ export class Messaging {
                         this.createMessage(body, message, thread);
                     }
                     break;
+                case "mail.record/insert":
```
Comment (phenix-factory) [inline addons/mail/static/src/messaging.js]: @seb-odoo @alexkuhn should we keep this notification name ? Doesn't make a lot of sens in the new code.

Comment (alexkuhn) [reply addons/mail/static/src/messaging.js]: There's no reason to remove, and actually there are very good design choices with a single notification format, so would keep it for the moment.

File: addons/mail/static/src/discuss/partner_im_status.xml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    <t t-name="mail.partner_im_status" owl="1">
+        <span class="o-mail-partner-im-status d-flex flex-column fa-stack">
+            <t t-if="props.partner.im_status === 'online'">
+                <i class="o_PersonaImStatusIconView_icon o-online fa fa-circle fa-stack-1x text-primary" title="Online" role="img" aria-label="User is online"/>
+            </t>
+            <t t-if="props.partner.im_status === 'away'">
+                <i class="o_PersonaImStatusIconView_icon o-away fa fa-circle fa-stack-1x text-warning" title="Idle" role="img" aria-label="User is idle"/>
+            </t>
+            <t t-if="props.partner.im_status === 'offline'">
+                <i class="o_PersonaImStatusIconView_icon o-offline fa fa-circle-o fa-stack-1x text-700" title="Offline" role="img" aria-label="User is offline"/>
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/partner_im_status.xml]: `.o_PersonaImStatusIconView_icon` => `.o-mail-partner-im-status-icon`

File: addons/mail/static/src/discuss/partner_im_status.xml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+            </t>
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/partner_im_status.xml]: All the `t-if` can be attribute of `<i>` themselves. Would make template more compact.

File: addons/mail/static/src/discuss/partner_im_status.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+/** @odoo-module **/
+
+import { Component } from "@odoo/owl";
+
+export class PartnerImStatus extends Component {
+    static template = "mail.partner_im_status";
+    static props = ["*"];
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/partner_im_status.js]: should define explicitly the props instead. I see notably `partner` that looks mandatory

File: addons/mail/static/src/discuss/sidebar.xml [hunk 7]
```diff
@@ -105,8 +105,11 @@
         }"
         t-on-click="() => this.openThread(threadId)"
     >
-        <div class="position-relative d-flex ms-4" style="width:30px;height:30px">
+        <div class="position-relative d-flex ms-4" style="width:30px;height:30px" t-on-click="() => this.messaging.openDiscussion(thread.id)">
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/sidebar.xml]: This `t-on-click` looks useless.
There's already a `t-on-click` on the parent above to `openThread(threadId)`, so it doesn't functionally the same.

File: addons/mail/static/src/messaging.js [hunk 8]
```diff
@@ -243,9 +254,10 @@ export class Messaging {
         if (id in this.partners) {
             return this.partners[id];
         }
-        const partner = { id, name };
+        const partner = { id, name, im_status: "" };
         this.partners[id] = partner;
-        return partner;
+        this.updateImStatusRegistration();
+        return reactive(partner);
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging.js]: You should not return `reactive()` here. All `Messaging` state is a reactive already (see `MessagingService`)

File: addons/mail/static/src/messaging.js [hunk 9]
```diff
@@ -263,6 +275,15 @@ export class Messaging {
                         this.createMessage(body, message, thread);
                     }
                     break;
+                case "mail.record/insert":
+                    {
+                        const partners_in_payload = notif.payload.Partner;
+                        for (const partner in partners_in_payload) {
+                            this.partners[partners_in_payload[partner].id] =
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging.js]: Certainly need to `Object.assign()` rather than fully replace

File: addons/mail/static/src/messaging.js [hunk 10]
```diff
@@ -132,6 +133,16 @@ export class Messaging {
...
 
+    updateImStatusRegistration() {
+        const partnerIds = [];
+        for (const partner in this.partners) {
+            if (partner.im_status !== "im_partner" && !partner.is_public) {
+                partnerIds.push(partner);
+            }
+        }
+        this.imStatusService.registerToImStatus("res.partner", partnerIds);
+    }
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging.js]: Note that this code is quite inefficient. It's called whenever a partner is created, but it has a for-loop on all partners.
I would be best to add/remove the single partner in the collection.

File: addons/mail/static/src/messaging_menu/messaging_menu.xml [hunk 11]
```diff
@@ -30,8 +30,11 @@
 <t t-name="mail.messaging_menu.preview" owl="1">
     <t t-set="thread" t-value="messaging.threads[preview.id]"/>
     <div class="o-mail-notification-item d-flex cursor-pointer border-bottom align-items-center" t-on-click="() => this.openDiscussion(thread.id)">
-        <div class="m-1" style="width:40px;height:40px;">
+        <div class="position-relative d-flex m-1" style="width:40px;height:40px;">
             <img class="w-100 h-100 rounded-circle" alt="Thread Image" t-att-src="thread.imgUrl" />
+            <div class="d-flex justify-content-center flex-shrink-0 position-absolute bottom-0 end-0 align-items-center rounded-circle bg-100 small o-muted">
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging_menu/messaging_menu.xml]: Note that background is still hardcoded color. So it doesn't properly change in some contexts.

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/messaging.js (hunk 9 above)
Comment (maki-odoo) [reply addons/mail/static/src/messaging.js]: It is just im-status that is needed. So instead of Object.assign I just replace the im_status property if It is OK.

File: addons/mail/static/src/messaging_menu/messaging_menu.xml (hunk 11 above)
Comment (maki-odoo) [reply addons/mail/static/src/messaging_menu/messaging_menu.xml]: I see and I put a TODO in related scss to work on it ASAP.

File: addons/mail/static/src/messaging.js (hunk 9 above)
Comment (alexkuhn) [reply addons/mail/static/src/messaging.js]: Ok for now :)

File: addons/mail/static/src/messaging.js (hunk 10 above)
Comment (maki-odoo) [reply addons/mail/static/src/messaging.js:144]: The partnerIds in the registerToImStatus is a list. I can make it trigger only when we pass other extra argument to the createPartner, for example at the end of a loop of partners creation. What is the better way?

Comment (alexkuhn) [reply addons/mail/static/src/messaging.js]: You can store locally `registeredImStatusPartners`, and just add/remove whenever you need to update it. For example, when creating a partner, it should only add the new `partnerId` in the list and call `registerToImStatus` again.

This avoids an unnecessary for-loop, which is costly if you have a lot of partners.

Comment (maki-odoo) [reply addons/mail/static/src/messaging.js]: You mean I have to add the new partner to the existing partnerIds list in the imStatusService? and then where should I delete it?

File: addons/mail/static/src/discuss/partner_im_status.xml [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+    <t t-name="mail.partner_im_status" owl="1">
+        <span t-if="props.partner.im_status" class="o-mail-partner-im-status d-flex flex-column fa-stack">
+            <i t-if="props.partner.im_status === 'online'" class="o-mail-partner-im-status-icon o-online fa fa-circle fa-stack-1x text-primary" title="Online" role="img" aria-label="User is online" />
+            <i t-if="props.partner.im_status === 'away'" class="o-mail-partner-im-status-icon o-away fa fa-circle fa-stack-1x text-warning" title="Idle" role="img" aria-label="User is idle" />
+            <i t-if="props.partner.im_status === 'offline'" class="o-mail-partner-im-status-icon o-offline fa fa-circle-o fa-stack-1x text-700" title="Offline" role="img" aria-label="User is offline" />
+            <i t-if="props.partner.im_status === 'bot'" class="o-mail-partner-im-status-icon o-bot fa fa-heart fa-stack-1x text-primary" title="Bot" role="img" aria-label="User is a bot"/>
+        </span>
+    </t>
+
+</templates>
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/partner_im_status.xml]: missing newline at End Of Line

File: addons/mail/static/src/messaging.js [hunk 13]
```diff
@@ -36,6 +37,7 @@ export class Messaging {
         this.threads = {};
         this.users = {};
         this.internalUserGroupId = null;
+        this.registeredImStatusPartners = markRaw([]);
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging.js]: Why you need `markRaw` here?
If a component observes it, then should be reactive too.

File: addons/mail/static/src/messaging_service.js [hunk 14]
```diff
@@ -16,7 +16,7 @@ export const messagingService = {
             threadId = parseInt(activeId.slice(13), 10);
         }
 
-        const messaging = new Messaging(env, rpc, orm, user, router, threadId, notification);
+        const messaging = new Messaging(env, rpc, orm, user, router, threadId, im_status, notification);
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging_service.js]: prettier tells LOC is too long, and should be reformated

File: addons/mail/static/src/messaging.js [hunk 15]
```diff
@@ -243,8 +252,9 @@ export class Messaging {
         if (id in this.partners) {
             return this.partners[id];
         }
-        const partner = { id, name };
+        const partner = { id, name, im_status: "" };
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging.js]: can set `im_status` to `null` here

File: addons/mail/static/src/messaging.js [hunk 16]
```diff
@@ -263,6 +273,15 @@ export class Messaging {
                         this.createMessage(body, message, thread);
                     }
                     break;
+                case "mail.record/insert":
+                    {
+                        const partners_in_payload = notif.payload.Partner;
+                        for (const partner in partners_in_payload) {
+                            this.partners[partners_in_payload[partner].id].im_status =
+                                partners_in_payload[partner].im_status;
+                        }
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging.js]: Would be easier to read these LOCs with a more brief local variable. Even just `data` is good (it's local variable anyway, so context is easy to grasp)

File: addons/mail/static/src/messaging.js [hunk 17]
```diff
@@ -263,6 +273,15 @@ export class Messaging {
                         this.createMessage(body, message, thread);
                     }
                     break;
+                case "mail.record/insert":
+                    {
+                        const partners_in_payload = notif.payload.Partner;
+                        for (const partner in partners_in_payload) {
```
Comment (alexkuhn) [inline addons/mail/static/src/messaging.js]: the `in` iterable is `partnerId`, not `partner`. Renaming accordingly will improve readability, and even simplifying code (`partners_in_payload[partner].id` is just `partnerId`)

File: addons/mail/static/src/discuss/partner_im_status.scss [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+
+.o-mail-partner-im-status {
+    width: $o-mail-partner-status-icon-size;
+    height: $o-mail-partner-status-icon-size;
+    line-height: 1.3em;
+}
+
+.o-mail-category-im-status {
+    transform: scale(0.75);
+}
```
Comment (alexkuhn) [inline addons/mail/static/src/discuss/partner_im_status.scss]: This style should be in `sidebar.scss`.
This is a bit annoying to deal with proper scaling based on avatar scaling...

I have a feeling we'll need to revamp this code eventually, so that we won't have to manage scaling by hand.
But ok to keep it like this for the time being. Acceptable increase of technical debt: having the feature is more important for the moment.

File: addons/mail/static/src/thread/message.xml [hunk 19]
```diff
@@ -7,6 +7,9 @@
             <div t-if="!props.squashed" class="o-mail-avatar-container position-relative">
                 <img class="w-100 h-100 rounded-circle o_object_fit_cover o_redirect" t-att-class="{ 'cursor-pointer': props.hasActions }"
                     t-attf-src="/mail/channel/1/partner/{{message.authorId}}/avatar_128"/>
+                <div class="d-flex justify-content-center flex-shrink-0 position-absolute bottom-0 end-0 align-items-center rounded-circle bg-100 small o-muted">
+                    <PartnerImStatus partner="author"/> <!-- TODO openchat func associated with partner -->
```
Comment (alexkuhn) [inline addons/mail/static/src/thread/message.xml]: Remove comment.
It's preferable to have such comment in pad somewhere else, rather than in code.

Also this feature is not on the im status icon itself, but the avatar. It just happens in master we have different handlers for image and im status icon, but it maps to the same feature (opening chat of corresponding partner).
Ideally a click on avatar opens chat. The IM status icon should be part of the avatar.

But ok for the mean time to not introduce such component that combines im status icon and avatar.
Will do in follow-up PR.

File: addons/mail/static/src/messaging.js (hunk 13 above)
Comment (maki-odoo) [reply addons/mail/static/src/messaging.js]: I needed a list not proxy to pass to registerToImStatus function. I moved markRaw to  updateImStatusRegistration function.

**Comment by alexkuhn** [comment]
Same remark as this comment in another PR:
https://github.com/odoo-dev/odoo/pull/1410#issuecomment-1317772954

**tl;dr:** history of target branch changed, need to cherry-pick and put code in `mail/static/src/new/` folder.

**Comment by maki-odoo** [comment]
> Same remark as this comment in another PR: [#1410 (comment)](https://github.com/odoo-dev/odoo/pull/1410#issuecomment-1317772954)
> 
> **tl;dr:** history of target branch changed, need to cherry-pick and put code in `mail/static/src/new/` folder.

File: addons/mail/static/src/new/discuss/sidebar.xml [hunk 20]
```diff
@@ -107,6 +107,9 @@
     >
         <div class="position-relative d-flex ms-4" style="width:30px;height:30px">
             <img class="w-100 h-100 rounded-circle" t-att-src="thread.imgUrl" alt="Thread Image"/>
+            <div class="o-mail-category-im-status d-flex justify-content-center flex-shrink-0 position-absolute bottom-0 end-0 align-items-center rounded-circle bg-100 small o-muted">
```
Comment (alexkuhn) [inline addons/mail/static/src/new/discuss/sidebar.xml]: `.o-muted` looks useless

File: addons/mail/static/src/new/messaging_menu/messaging_menu.xml [hunk 21]
```diff
@@ -30,8 +30,11 @@
 <t t-name="mail.messaging_menu.preview" owl="1">
     <t t-set="thread" t-value="messaging.threads[preview.id]"/>
     <div class="o-mail-notification-item d-flex cursor-pointer border-bottom align-items-center" t-on-click="() => this.openDiscussion(thread.id)">
-        <div class="m-1" style="width:40px;height:40px;">
+        <div class="position-relative d-flex m-1" style="width:40px;height:40px;">
             <img class="w-100 h-100 rounded-circle" alt="Thread Image" t-att-src="thread.imgUrl" />
+            <div class="d-flex justify-content-center flex-shrink-0 position-absolute bottom-0 end-0 align-items-center rounded-circle bg-100 small o-muted">
```
Comment (alexkuhn) [inline addons/mail/static/src/new/messaging_menu/messaging_menu.xml]: `.o-muted` looks useless

File: addons/mail/static/src/new/thread/message.xml [hunk 22]
```diff
@@ -7,6 +7,9 @@
             <div t-if="!props.squashed" class="o-mail-avatar-container position-relative">
                 <img class="w-100 h-100 rounded-circle o_object_fit_cover o_redirect" t-att-class="{ 'cursor-pointer': props.hasActions }"
                     t-attf-src="/mail/channel/1/partner/{{message.authorId}}/avatar_128"/>
+                <div class="d-flex justify-content-center flex-shrink-0 position-absolute bottom-0 end-0 align-items-center rounded-circle bg-100 small o-muted">
```
Comment (alexkuhn) [inline addons/mail/static/src/new/thread/message.xml]: `.o-muted` looks useless

File: addons/mail/static/src/new/messaging.js [hunk 23]
```diff
@@ -276,6 +289,14 @@ export class Messaging {
                         this.createMessage(body, message, thread);
                     }
                     break;
+                case "mail.record/insert":
+                    {
+                        const partnersInPayload = notif.payload.Partner;
+                        partnersInPayload.forEach(({ id, im_status }) => {
+                            this.partners[id].im_status = im_status;
+                        });
```
Comment (alexkuhn) [inline addons/mail/static/src/new/messaging.js]: Prefer `for () {}` syntax rather than `.forEach()`, for improved readability

File: addons/mail/static/src/models/messaging.js [hunk 24]
```diff
@@ -259,7 +259,9 @@ Patch({
             for (const guest of this.models["Guest"].all()) {
                 guestIds.push(guest.id);
             }
+            /* @maki: It passes the partner with im_status = im_partner to imStatusService in the new mail code.
             this.env.services["im_status"].registerToImStatus("res.partner", partnerIds);
+*/
```
Comment (alexkuhn) [inline addons/mail/static/src/models/messaging.js]: Ok to comment this LOC. No need to put a comment: just disable the feature.

File: addons/mail/static/src/new/messaging_menu/messaging_menu.xml [hunk 25]
```diff
@@ -30,8 +30,11 @@
...
+                <PartnerImStatus t-if="thread.type === 'chat'" partner="messaging.partners[thread.chatPartnerId]"/>
```
Comment (alexkuhn) [inline addons/mail/static/src/new/messaging_menu/messaging_menu.xml]: The `t-if` should be moved to `div` above, as you don't want to show `rounded-circle` neither when there's no `im_status` to show

File: addons/mail/static/src/new/thread/message.xml (hunk 22 above)
Comment (alexkuhn) [inline addons/mail/static/src/new/thread/message.xml]: This wrapping `<div>` is always present when using `PartnerImStatus`, so makes sense to have it part of the component. That way we should not manually maintain in each context, and also makes it easier to reuse the component in some other contexts.

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/new/messaging.js (hunk 23 above)
Comment (maki-odoo) [reply addons/mail/static/src/new/messaging.js]: It is not itterable and `for of` doesn't work on it and `for in` is kind of bulky. Do you want me to change it anyway?

Comment (alexkuhn) [reply addons/mail/static/src/new/messaging.js]: From what I see, this is an array for data related to `im_status`.

It's just that the notification name is quite generic, so it can be an object of an array.
It just so happens that fetched `im_status` are returned as an array for `Partner`.
For other notifications `mail.record/insert`, `notif.payload.Partner` can be `undefined`, or even an object.

So you should make sure that `notif.payload.Partner` exists, and if so you must handle case of a single object or array of objects.

Also if the payload has no `im_status`, you should not update `im_status`. You should also make sure of that in this LOC.

File: addons/mail/static/src/new/messaging.js [hunk 26]
```diff
@@ -395,7 +430,7 @@ export class Messaging {
                 .call("mail.channel", "channel_fetch_preview", [ids])
                 .then((previews) => {
                     for (const preview of previews) {
-                        preview.last_message.date = markRaw(
+                        preview.last_message.date = toRaw(
```
Comment (alexkuhn) [inline addons/mail/static/src/new/messaging.js]: This should stay `markRaw`, as it's stored in state.

File: addons/mail/static/src/new/messaging.js [hunk 27]
```diff
@@ -134,6 +136,16 @@ export class Messaging {
         });
     }
 
+    updateImStatusRegistration(partner) {
+        if (partner.im_status !== "im_partner" && !partner.is_public) {
+            this.registeredImStatusPartners.push(partner.id);
+        }
+        this.imStatusService.registerToImStatus(
+            "res.partner",
+            markRaw(this.registeredImStatusPartners)
```
Comment (alexkuhn) [inline addons/mail/static/src/new/messaging.js]: This should be `toRaw()`, as it's passed to external API (it's stored as a reactive, and should be kept as so)

File: addons/mail/static/src/models/messaging.js [hunk 28]
```diff
@@ -259,7 +259,7 @@ Patch({
             for (const guest of this.models["Guest"].all()) {
                 guestIds.push(guest.id);
             }
-            this.env.services["im_status"].registerToImStatus("res.partner", partnerIds);
+            //            this.env.services["im_status"].registerToImStatus("res.partner", partnerIds);
```
Comment (alexkuhn) [inline addons/mail/static/src/models/messaging.js]: Ok for comment, just keep single whitespace between `//` and beginning of the statement.

File: addons/mail/static/src/new/messaging_menu/messaging_menu.xml [hunk 29]
```diff
@@ -30,8 +30,9 @@
 <t t-name="mail.messaging_menu.preview" owl="1">
     <t t-set="thread" t-value="messaging.threads[preview.id]"/>
     <div class="o-mail-notification-item d-flex cursor-pointer border-bottom align-items-center" t-on-click="() => this.openDiscussion(thread.id)">
-        <div class="m-1" style="width:40px;height:40px;">
+        <div class="position-relative d-flex m-1" style="width:40px;height:40px;">
             <img class="w-100 h-100 rounded-circle" alt="Thread Image" t-att-src="thread.imgUrl" />
+            <PartnerImStatus t-if="thread.type === 'chat'" partner="messaging.partners[thread.chatPartnerId]"/>
```
Comment (alexkuhn) [inline addons/mail/static/src/new/messaging_menu/messaging_menu.xml]: Ok for the moment to do this: manually add img and `PartnerImStatus` (even if partner has no im status, component guards it).

Note that this code doesn't properly link `thread.imgUrl` with `thread.chatPartnerId`, hence why you need some code that we should actively maintain to sync data properly.

It's ok for the moment, but eventually we should do something, because this adds some significant technical debts 😅

File: addons/mail/static/src/new/messaging.js (hunk 26 above)
Comment (maki-odoo) [reply addons/mail/static/src/new/messaging.js]: Sorry, I changed the wrong one!

File: addons/mail/static/src/new/discuss/sidebar.xml [hunk 30]
```diff
@@ -107,6 +107,7 @@
     >
         <div class="position-relative d-flex ms-4" style="width:30px;height:30px">
             <img class="w-100 h-100 rounded-circle" t-att-src="thread.imgUrl" alt="Thread Image"/>
+            <PartnerImStatus t-if="thread.type === 'chat'" className="'o-mail-category-im-status'" partner="messaging.partners[thread.chatPartnerId]"/>
```
Comment (alexkuhn) [inline addons/mail/static/src/new/discuss/sidebar.xml]: To bad all these `position-relative` that come from `PartnerImStatus` using `position-absolute` internally.
Right now I know why we have them, but it might become hard to see why.

Could you move `position-absolute bottom-0 end-0` from `PartnerImStatus` to `props.className` instead?
It's a bit more diff, but at least we'd see the link between `position-relative` and corresponding `position-absolute`. Also would allow reusing `PartnerImStatus` in non-absolute positioning, like how it's done in `master` for partner mentions.

**Review by alexkuhn — COMMENTED** [review]
Small comment then good for merge!

File: addons/mail/static/src/new/discuss/sidebar.xml [hunk 31]
```diff
@@ -107,6 +107,7 @@
     >
         <div class="position-relative d-flex ms-4" style="width:30px;height:30px">
             <img class="w-100 h-100 rounded-circle" t-att-src="thread.imgUrl" alt="Thread Image"/>
+            <PartnerImStatus t-if="thread.type === 'chat'" className="'o-mail-category-im-status'" hasPositionAbsolute="true" partner="messaging.partners[thread.chatPartnerId]"/>
```
Comment (alexkuhn) [inline addons/mail/static/src/new/discuss/sidebar.xml]: Instead of introducing `hasPositionAbsolute` prop, it's best to give the classnames in `className` prop.

That way, no need to define another prop, and the style customisation of this component is apparent in this template. This is desirable so that we properly see `position-relative` and `position-absolute` at the same place.

**Review by alexkuhn — COMMENTED** [review]
LGTM otherwise

Comment about using `className` rather than introducing `hasPositionAbsolute` applies on all occurrences of `<PartnerImStatus>`.

Can you also make at least a simple test to show partner im status next to avatar, e.g. next to a message?

**Review by alexkuhn — CHANGES_REQUESTED** [review]
Please look at re-enabling the following tests:

- "sidebar: chat im_status rendering"
- all tests in `persona_im_status_icon_tests.js`

That way there's a few test coverage on this new code.

**Comment by maki-odoo** [comment]
> Please look at re-enabling the following tests:
> 
> * "sidebar: chat im_status rendering"
> * all tests in `persona_im_status_icon_tests.js`
> 
> That way there's a few test coverage on this new code.

Done

**Review by alexkuhn — CHANGES_REQUESTED** [review]
👆rebased + fixed small issue that kept some tests freezing due to not deploying `imStatusService` for `TestServer` tests.
There are still 15 tests failing in this branch

**Comment by maki-odoo** [comment]
> point_up_2rebased + fixed small issue that kept some tests freezing due to not deploying `imStatusService` for `TestServer` tests. There are still 15 tests failing in this branch

Done
